apps/aecore/src/aec_sync.erl

```diff
@@ -230,132 +203,198 @@ terminate(_, _) ->
+        false    -> ok;
+        {_, Old} -> lager:info("Peer ~p already has a worker (~p)", [ppp(PeerId), Old])
+    end,
+    erlang:monitor(process, Pid),
+    lager:debug("New worker ~p for peer ~p", [Pid, ppp(PeerId)]),
+    ST#sync_task{ workers = lists:keystore(PeerId, 1, Ws, {PeerId, Pid}) };
+do_handle_worker({change_worker, PeerId, OldPid, NewPid}, ST = #sync_task{ workers = Ws }) ->
+    case lists:keyfind(PeerId, 1, Ws) of
+        false ->
+            lager:info("Missing worker ~p for peer ~p", [OldPid, ppp(PeerId)]);
+        {_, OldPid} ->
+            ok;
+        {_, AnotherPid} ->
+            lager:info("Wrong worker stored for peer ~p (~p)", [ppp(PeerId), AnotherPid])
+    end,
+    erlang:monitor(process, NewPid),
+    erlang:demonitor(OldPid, [flush]),
+    lager:debug("Update worker ~p (was ~p) for peer ~p", [NewPid, OldPid, ppp(PeerId)]),
+    ST#sync_task{ workers = lists:keystore(PeerId, 1, Ws, {PeerId, NewPid}) }.
+
+do_terminate_worker(Pid, S = #state{ sync_tasks = STs }) ->
+    case [ ST || ST <- STs, lists:keyfind(Pid, 2, ST#sync_task.workers) /= false ] of
+        [ST] ->
+            set_sync_task(do_terminate_worker(Pid, ST), S);
+        [] ->
+            S
+    end;
+do_terminate_worker(Pid, ST = #sync_task{ workers = Ws }) ->
+    {Peer, _} = lists:keyfind(Pid, 2, Ws),
+    lager:debug("Terminating worker ~p for peer ~p", [Pid, ppp(Peer)]),
+    ST#sync_task{ workers = lists:keydelete(Pid, 2, Ws) }.
 
 %%%=============================================================================
 %%% Jobs worker
 %%%=============================================================================
-
-sync_worker() ->
-    Res = jobs:dequeue(sync_jobs, 1),
-    lager:debug("dequed job ~p", [pp(Res)]),
-    process_job(Res).
-
-%% Note: we always dequeue exactly ONE job
-process_job([{_T, Job}]) ->
-    case Job of
-        {forward, #{block := Block}, PeerId} ->
-            do_forward_block(Block, PeerId);
-        {forward, #{tx := Tx}, PeerId} ->
-            do_forward_tx(Tx, PeerId);
-        {start_sync, PeerId, RemoteHash} ->
-            case sync_in_progress(PeerId) of
-              false -> do_start_sync(PeerId, RemoteHash);
-              _     -> lager:info("sync already in progress ~p", [ppp(PeerId)])
-            end;
-        {fetch_mempool, PeerId} ->
-            do_fetch_mempool(PeerId);
-        {ping, PeerId} ->
-            ping_peer(PeerId);
-        _Other ->
-            lager:debug("unknown job", [])
-    end.
-
-enqueue(Op, Msg) ->
-    Peers = aec_peers:get_random(all),
-    [ jobs:enqueue(sync_jobs, {Op, Msg, aec_peers:peer_id(PeerInfo)})
-      || PeerInfo <- Peers ].
+delayed_run_job(PeerId, Task, Queue, Fun, Delay) ->
+    OldWorker = self(),
+    spawn(fun() -> NewWorker = self(),
+                   handle_worker(Task, {change_worker, PeerId, OldWorker, NewWorker}),
+                   timer:sleep(Delay),
+                   jobs:run(Queue, Fun)
+          end).
+
+run_job(Queue, Fun) ->
```

Comment:
I cannot really oversee what the difference is between using the jobs package and spawning workers.
Do we possibly loose anything in this way, such as load regulation, zombie process cleanup, things like that? You still call jobs, but via an extra spawn?